test_conformance/api/negative_device.cpp

```diff
@@ -0,0 +1,509 @@
+    {
+        err = clGetPlatformInfo(platform, CL_PLATFORM_HOST_TIMER_RESOLUTION,
+                                sizeof(timer_resolution), &timer_resolution,
+                                nullptr);
+        test_error(err, "clGetPlatformInfo");
+    }
+
+    if (timer_resolution != 0)
+    {
+        log_info("Platform Supports Timers\n");
+        log_info("Skipping CL_INVALID_OPERATION tests\n");
+
+        err = clGetHostTimer(deviceID, nullptr);
+        test_failure_error_ret(err, CL_INVALID_VALUE,
+                               "clGetHostTimer should return CL_INVALID_VALUE "
+                               "when: \"host_timestamp is NULL\"",
+                               TEST_FAIL);
+    }
+    else
+    {
+        log_info("Platform does not Support Timers\n");
+        log_info("Skipping CL_INVALID_VALUE tests\n");
+
+        err = clGetHostTimer(deviceID, host_timestamp);
+        test_failure_error_ret(
+            err, CL_INVALID_OPERATION,
+            "clGetHostTimer should return CL_INVALID_OPERATION when: \"the "
+            "platform associated with device does not support device and host "
+            "timer synchronization\"",
+            TEST_FAIL);
+    }
+
+    return TEST_PASS;
+}
+
+/* Negative Tests for clCreateSubDevices */
+enum SupportedPartitionSchemes
+{
+    None = 0,
+    Equally = 1 << 0,
+    Counts = 1 << 1,
+    Affinity = 1 << 2,
+    All_Schemes = Affinity | Counts | Equally,
+};
+
+static int get_supported_properties(cl_device_id deviceID)
+{
+    size_t number_of_properties = 0;
+    int err = clGetDeviceInfo(deviceID, CL_DEVICE_PARTITION_PROPERTIES, 0,
+                              nullptr, &number_of_properties);
+    test_error(err, "clGetDeviceInfo");
+    std::vector<cl_device_partition_property> supported_properties(
+        number_of_properties / sizeof(cl_device_partition_property));
+    err = clGetDeviceInfo(deviceID, CL_DEVICE_PARTITION_PROPERTIES,
+                          number_of_properties, &supported_properties.front(),
+                          nullptr);
+    test_error(err, "clGetDeviceInfo");
+    int ret = SupportedPartitionSchemes::None;
+    for (auto property : supported_properties)
+    {
+        switch (property)
+        {
+            case CL_DEVICE_PARTITION_EQUALLY:
+                ret |= SupportedPartitionSchemes::Equally;
+                break;
+            case CL_DEVICE_PARTITION_BY_COUNTS:
+                ret |= SupportedPartitionSchemes::Counts;
+                break;
+            case CL_DEVICE_PARTITION_BY_AFFINITY_DOMAIN:
+                ret |= SupportedPartitionSchemes::Affinity;
+                break;
+            default: break;
+        }
+    }
+    return ret;
+}
+
+static std::vector<cl_device_partition_property>
+get_invalid_properties(int unsupported_properties)
+{
+    if (unsupported_properties & SupportedPartitionSchemes::Equally)
+    {
+        return { CL_DEVICE_PARTITION_EQUALLY, 1, 0 };
```

Comment:
Not sure this is worth worrying about, but there is a chance partitioning with these properties could exceed `CL_DEVICE_PARTITION_MAX_SUB_DEVICES`.

Reply:
This function is used to create a set of properties which the implementation does not support. The values in here should be "valid" but for a given implementation should never succeed.

Reply:
OK, I see, these properties should only be used when an implementation does not support `CL_DEVICE_PARTITION_EQUALLY`, to test that `CL_INVALID_VALUE` is returned.

Theoretically I think an implementation could return `CL_DEVICE_INVALID_PARTITION_COUNT` here, if the total number of sub-devices when partitioning equally exceeds the maximum number of sub-devices, but in practices I bet most implementations would check that this partitioning type is supported and hence would return `CL_INVALID_VALUE` first.

I don't think this is worth worrying about.